V1.4 fix inconsitent error messages for invalid index

nbvannhi opened the pull request:
Closes #223 
Closes #217 

`INDEX` error message summary
- If `INDEX` is specified but command is invalid,
_invalid command format_ error message will be shown with correct command format.
- If `INDEX` is unable to parse,
_invalid index_ error message will be shown (including both negative and exceeding `MAX_INT` cases).
- If `INDEX` exceeds no. students,
_invalid student index_ error message will be thrown with no. students.

Note: `edit 1 some_random_texts` will shown _invalid index_ error message instead of _invalid command format_ error message.

nbvannhi committed:
Fix inconsistent error messages for invalid index

jonas-chow commented on src/main/java/seedu/address/logic/commands/ShowCommand.java:
```diff
@@ -121,7 +121,7 @@ private CommandResult showStudentByIndex(Model model) throws CommandException {
         List<Student> students = model.getFilteredStudentList();
 
-        if (index.getZeroBased() >= students.size()) {
+        if (index.getOneBased() < 1 || index.getZeroBased() >= students.size()) {
```
The added clause will never be true. `ParseIndex` takes care of this case already.

jonas-chow commented on src/main/java/seedu/address/logic/parser/ShowCommandParser.java:
```diff
@@ -47,7 +48,7 @@ public ShowCommand parseByIndex(ArgumentMultimap argMultimap, Path savePath) thr
             index = ParserUtil.parseIndex(argMultimap.getPreamble());
         } catch (ParseException pe) {
-            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, ShowCommand.MESSAGE_USAGE));
+            throw new ParseException(MESSAGE_INVALID_STUDENT_DISPLAYED_INDEX);
```
Perhaps we need a more descriptive error message than "The student index provided is invalid".
I suggest we split the error message into the following cases:

* Inputs that contain characters that are not digits: "Index is not a positive unsigned integer."
* Inputs that only contain digits
    * Those that `parseInt` properly: 
        * If 0: "Index is not a positive unsigned integer."
        * If greater than student list size: "Index is too large"
        * Else: Executes properly
    * Doesn't `parseInt`: Above `MAX_INT`: "Index is too large"

Would be nice if you could rewrite the `parseIndex` function in `ParserUtil.java` to implement these (except perhaps the greater than student list size one)

If it's in the `parseIndex` function, then it'd be general enough to also affect the other related commands like `edit` and `delete`.

zhing22 commented:
yes could you please help to fix all other commands that uses index? (namely show, delete, and edit)

also since you are working on it could you take over #217 too? its about error message for negative numbers and 0.

Thank you!! <3

jonas-chow reviewed (commented):
I feel like the error message is too general here.
Things like `show a` and `show -1` will give the same error message, but the error message doesn't really tell the user how they should be changing their input to make it valid.

nbvannhi committed:
Merge branch 'master' of https://github.com/AY2122S1-CS2103T-W08-2/tp into v1.4-fix-show

nbvannhi committed:
Fix error when prefixes are specified in `show INDEX`

`show INDEX -a Midterm` doesn't throw an error as expected.

This is because ShowCommandParser::parseByIndex doesn't check if
there are any prefixes after `INDEX`.

nbvannhi committed:
Rename utility method

nbvannhi committed:
Refine ShowCommandParser::parse

nbvannhi committed:
Make `INDEX` error messages more descriptive

In summary,
- If `INDEX` is specified but command is invalid, invalid command
format error message will be shown with correct command format.
- If `show INDEX` fails to parse `INDEX`, invalid index error
message will be shown (including both negative and exceeding
MAX_INT cases).
- If `INDEX` exceeds no. students, invalid student index error
message will be thrown with no. students.

nbvannhi committed:
Fix unspecified INDEX error

`show` should throw invalid command format error

nbvannhi committed:
Refine ShowCommandParser

nbvannhi committed:
Edit `edit` and `delete` command to behave consistently with `show`

Summary of behavior
- If `INDEX` is specified but command is invalid, invalid command
format error message will be shown with correct command format.
- If `INDEX` is unable to parse, invalid index error message will
be shown (including both negative and exceeding MAX_INT cases).
- If `INDEX` exceeds no. students, invalid **student** index error
message will be shown **with no. students**.

nbvannhi committed:
Fix checkstyle

nbvannhi committed:
Merge branch 'master' into v1.4-fix-show

nbvannhi committed:
Remove trailing space

nbvannhi committed:
Fix checkstyle

nbvannhi committed:
Adapt LogicManagerTest to new error messages

nbvannhi committed:
Adapt CommandParserTest to new error messages

nbvannhi committed:
Adapt CommandParserTest to new error message for `edit` and `delete`

jonas-chow reviewed (changes requested):
`show 1` doesn't work!!!

jonas-chow committed:
Fix logical error in ShowCommandParser

jonas-chow reviewed (approved):
There was an issue with show command not working with indices. I fixed it by negating a condition. Otherwise, lgtm.

nbvannhi commented:
> 
> 
> `show 1` doesn't work!!!

Oh I fixed it later but forgot to push... Tks anw!
